blockchain/state_manager/src/lib.rs

```diff
@@ -375,6 +391,53 @@ where
         })
     }
 
+    fn tipset_executed_message(
+        block_store: &DB,
+        tipset: &Tipset,
+        cid: &Cid,
+        (message_from_address, message_sequence): (&Address, &u64),
+    ) -> Result<Option<MessageReceipt>, Error>
+    where
+        DB: BlockStore,
+    {
+        if tipset.epoch() == 0 {
+            return Ok(None);
+        }
+        let tipset = chain::tipset_from_keys(block_store, tipset.parents())
+            .map_err(|err| Error::Other(err.to_string()))?;
+        let messages = chain::messages_for_tipset(block_store, &tipset)
+            .map_err(|err| Error::Other(err.to_string()))?;
+        messages
+            .iter()
+            .enumerate()
+            .rev()
```

Comment:
Enumerate reverses indexes : https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=6688534596b324c37056a8927304d8ab

The method is traversing using reverse indexes : https://github.com/filecoin-project/lotus/blob/8c516df87b9f00c53c0767c607f955a7bab9d003/chain/stmgr/stmgr.go#L683

Reply:
```suggestion
            .rev()
            .enumerate()
```
Then this should be correct

Reply:
Could you help me understand? The indexes being used are all reversed. So in essence you enumerate and have [(0,m1),(1,m2),(2,m3)] when you do  reverse you have [(2,m3),(2,m2),(1,m1)] all the indexes being used assume they will start with 2 since they are starting from a back. If do rev first I wil l have (0,m3) which I don't think matches the index that is being used in the lotus implementation unless I am reading this wrong

Reply:
ah, oops, I misread the Lotus impl there :)